src/utils/getConfigManager.ts

```diff
@@ -0,0 +1,91 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import ConfigSet from '../types/ConfigSet';
+import getOptions from './getOptions';
+import loadConfig from './loadConfig';
+import { getConfigPathCandidatesForFile } from './getConfigsForFile';
+import NormalizedPath from '../types/NormalizedPath';
+
+class ConfigManager {
+    private configSet: ConfigSet = null;
+    private partialDiscoveredConfigs: ConfigSet = {};
+    private discoveredPaths: Set<string> = new Set();
+
+    public get all(): ConfigSet {
+        if (this.configSet === null) {
+            this._getAllConfigs();
+        }
+        return this.configSet;
+    }
+
+    public partialConfigSetForPath(configSourcePath: NormalizedPath): ConfigSet {
+        const partialSet: ConfigSet = {};
+
+        if (this.configSet) {
+            for (let configPathCandidate of getConfigPathCandidatesForFile(configSourcePath)) {
+                if (this.configSet[configPathCandidate]) {
+                    partialSet[configPathCandidate] = this.configSet[configPathCandidate];
+                }
+            }
+        } else {
+            for (let configPathCandidate of getConfigPathCandidatesForFile(configSourcePath)) {
+                const configPathCandidateFull = path.join(configPathCandidate, 'fence.json');
+                if (this.discoveredPaths.has(configPathCandidateFull)) {
+                    const discoveredConfig = this.partialDiscoveredConfigs[configPathCandidate];
+                    if (discoveredConfig) {
+                        partialSet[configPathCandidateFull] = discoveredConfig;
+                    }
+                } else {
+                    try {
+                        const stat = fs.statSync(configPathCandidateFull);
```

Comment:
We might be able to back this with the getAllFiles() array instead of the filesystem. Future work for microptimizations?

Reply:
Resolving as future work